### PR TITLE
Use Realm `queries` to get Messages for Actions functions

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #495

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 1]
```diff
@@ -269,7 +269,7 @@ class ThreadListFragment : Fragment(), SwipeRefreshLayout.OnRefreshListener {
             SwipeAction.QUICKACTIONS_MENU -> {
                 safeNavigate(
                     ThreadListFragmentDirections.actionThreadListFragmentToThreadActionsBottomSheetDialog(
-                        messageUidToReplyTo = MessageController.getMessageUidToReplyTo(thread.messages),
+                        messageUidToReplyTo = MessageController.getMessageToReplyTo(thread.messages).uid,
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:272]: it should not be done here, but directly where it is needed, i.e. in the destination fragment.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt [hunk 2]
```diff
@@ -41,36 +41,38 @@ class MessageActionsBottomSheetDialog : ActionsBottomSheetDialog() {
         val threadUid = navigationArgs.threadUid
         val messageUid = navigationArgs.messageUid
 
-        messageActionsViewModel.threadAndMessage(threadUid, messageUid).observe(viewLifecycleOwner) { (thread, message) ->
+        messageActionsViewModel.threadAndMessage(threadUid, messageUid).observe(viewLifecycleOwner) {
+
+            val (thread, message) = it
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt:46]: Why did you change it?

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt [hunk 3]
```diff
@@ -91,18 +88,55 @@ object MessageController {
...
 
-    fun getMessageUidToReplyTo(messages: RealmList<Message>): String = getMessageToReplyTo(messages).uid
+    private fun getMessagesAndDuplicates(thread: Thread, condition: String): List<Message> {
+
+        val messages = thread.messages.query(condition).find()
+
+        val byMessagesIds = "${Message::messageId.name} IN {${messages.map { it.messageId }.joinToString { "'$it'" }}}"
+        val duplicates = thread.duplicates.query(byMessagesIds)
+
+        return messages + duplicates.find()
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt:134]:
```suggestion
        val byMessagesIds = "${Message::messageId.name} IN {${messages.joinToString { "'${it.messageId}'" }}}"
        val duplicates = thread.duplicates.query(byMessagesIds).find()

        return messages + duplicates
```

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt [hunk 4]
```diff
@@ -91,18 +88,55 @@ object MessageController {
...
+        return getMessagesAndDuplicates(thread, "$isNotScheduled AND $isNotFromMe")
+    }
+
+    fun getDeletableMessages(thread: Thread): List<Message> {
+        return getMessagesAndDuplicates(thread, isNotScheduled)
+    }
+
+    fun getPermanentlyDeletableMessages(thread: Thread): List<Message> {
+        return thread.messages + thread.duplicates
+    }
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt:120]: It's not a query, you can move it to `Thread` class

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt [hunk 5]
```diff
@@ -91,18 +88,55 @@ object MessageController {
...
+        return messages.query("$isNotDraft AND $isNotFromMe").find().lastOrNull()
+            ?: messages.query(isNotDraft).find().lastOrNull()
+            ?: messages.last()
+    }
 
-        val isNotFromMe = "SUBQUERY(${Message::from.name}, \$recipient, " +
-                "\$recipient.${Recipient::email.name} != '${AccountUtils.currentMailboxEmail}').@count > 0"
+    fun getUnseenMessages(thread: Thread): List<Message> {
+        val isUnseen = "${Message::seen.name} == false"
+        return getMessagesAndDuplicates(thread, isUnseen)
```
Comment (sirambd) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt:98]:
```suggestion
        return getMessagesAndDuplicates(thread, "${Message::seen.name} == false")
```

**Review by sirambd — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt (hunk 4 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt:119]: I'd rather let it stay here, with the other similar functions. All actions functions that return a list of Messages to execute the action on, are here.

Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/MessageController.kt:119]: 🤔 i understand but i still find this case special, limit since it is used only once, put it directly where it is called

**Review by sirambd — APPROVED** [review]
